### ISSUE TITLE
Outlier identification

### ISSUE THREAD

**Issue by chrzyki** [lexibank/lexibank#102]
_From @tresoldi on November 19, 2017 18:10_

I've written some basic code, pure Python with no dependencies, to identify outliers in a list of word forms (code here: https://gist.github.com/tresoldi/20baeb36c1dade87e8221a35523d5cc3 ), which could be expanded into a proper word-model, with smoothing, etc. It has been helpful to identify corrections I've missed, such as English glosses with an unusual syntax (like between curly commas).

Do you think it would be helpful to write a proper word-model code (with character n-grams and so on) and "true" outlier identification (like Dixon's Q test) and proper smoothing, adding it to the `analysis` command of `lexibank` to generate (along with missing source or concepts) suggestions for form/segment review?

_Copied from original issue: glottobank/lexibank-data#344_

**Comment by chrzyki** [comment]
_From @LinguList on November 20, 2017 7:9_

What does it basically do? I am sorry, but I don't have the time to really inspect the code, and I don't know what you mean by outliers. You mean after segmentation, or pre-segmentation?

**Comment by chrzyki** [comment]
_From @LinguList on November 20, 2017 7:12_

Ah, I think I see what you mean. As lingpy is a dependency of lexibank, and as this code may be useful for the general checking of a wordlist, since this comes pre-orthoprofile, right? I'd suggest to even add it to lingpy and include it into the CLI. This would allow for a command like:

```
$ lingpy outliers -i wordlist
```

But you could similarly add it to lexibank in a similar form. What do you think, @xrotwang and @SimonGreenhill ?

**Comment by chrzyki** [comment]
_From @tresoldi on November 20, 2017 10:9_

Yes, that is what you inferred in the second reply. "Outlier" is probably not clearest of all terms, but I was using it to point to the fact that this absolutely not a "linguistic" analysis, only a (dumb) numeric one based in observations.

I was thinking of pre-segmentation (which is when I'm using it), but in theory it could be used for segments as well. In any case, the code in the gist was written as needed, without a clear plan, I would update to a "true" n-gram model before including it. Please consider this just an idea. :)

**Comment by chrzyki** [comment]
_From @LinguList on November 20, 2017 10:18_

It might be handy to make a first check, also to see if segmentation works in the end. Consider the following behaviour: you have a wordlist (in our cldf-like type, maybe tab-separated lingpy-format), you point to the column where the forms are, run the stats, and see whether there are some nice disgusting things.

**Comment by chrzyki** [comment]
_From @tresoldi on November 22, 2017 17:31_

I've updated my code and the small utility is now working and being useful (see below the output for the `bowernpny` dataset, still in development). I've put it in a gist, if anyone wants to use it https://gist.github.com/tresoldi/2b7f1e0a8a2804f305309b9c12d790f3

Here the output, it caught both an English gloss and a capitalized transcription:

tiago@vimercate:~/src/private$ python3 outlier/seqoutlier.py ~/src/lexibank-data/datasets/bowernpny/cldf/cognates.csv Form | head -n 20
('BOORRI-MUL', -11.868724322304748)
('MURDALANGINY', -11.863348562588262)
('KUDADJININY', -11.757519486334322)
('Gürrügümá', -11.749938834060362)
('NGAPAPPI', -11.648591623587155)
('dʒɤnɤgɾɪ', -11.478116814363315)
('ŋwiyá-ŋa1yuru-badá2 BREAD', -11.471807512938062)
('vδaywvnv', -11.437570303552498)
('BIRRID', -11.366652964636804)
('dʊɹʊmuʊndɪɹɪ', -11.309727075114662)
('nɤndɤ', -11.297346394450063)
('see ‘camp’', -11.217972267114687)
('not listed in other dialects', -11.194399904620557)
('ᶁeɲe:ɲ', -11.183360831037625)
('oŋkiːrdɐl', -11.181269189433603)
('NADJA', -11.140401781783192)
('əruoδ', -11.140401781783192)
('BAALUN', -11.059560993034141)
('ɪndʒɪnjɪ', -10.983952572102384)
('only specific', -10.966429060416962)

**Comment by chrzyki** [comment]
_From @LinguList on November 22, 2017 17:55_

Nice, I can really see something like this being nicely integrated either into lexibank or lingpy. It seems definitely useful when working on new datasets.

**Comment by chrzyki** [comment]
_From @SimonGreenhill on November 22, 2017 18:16_

Cool -- I like this a lot. Looks like it could be really useful.

**Comment by chrzyki** [comment]
_From @tresoldi on November 22, 2017 18:40_

I also tested it with lists of segments (both aligned and not); it works but it is easier to spot errors with the plain form.

I also wrote some quick and dirty statistical tests (a kind of Bayesian estimator and a Dixon Q outlier test), but honestly they don't seem really worth -- to just print a sorted list like I ended up doing seems to be enough.

**Comment by chrzyki** [comment]
_From @LinguList on November 22, 2017 19:45_

I'm thinking of some simple way to integrate this (I just had a few issues with another dataset). Imagine, for lingpy-2.7 (I think, lingpy is the right place, now that it supports cldf as well), to have some command:

```
$ lingpy check dataset
```
We'd assume basic wordlist form, normal excel data would not work (but that's anyway the first step that needs to be done: convert data from excel idiosyncratic format to something useful). And then, one could add additional routines. You might think of just opening an issue in lingpy on that, and we can add it to our milestone lingpy-2.7. It could involve simple testing for segmentation as well.

**Comment by chrzyki** [comment]
_From @tresoldi on November 28, 2017 0:49_

After struggling with Sun's Sino-Tibetan data, I followed what I think was a suggestion by @LinguList and wrote yet another quick & dirty statistical tool for dataset curation, exploring ideas for alternatives to the automatic cognate mapping. It is not really an improvement, but I think more and more that such approaches should be explored and more tools like that should be written. If you are interested, it is at https://gist.github.com/tresoldi/439c8af9d7c6c1f7bb8ea5ed06ea0bed and it uses a little function for sequence similarity I wrote (`seqsim` on pip) -- of course it can be replaced with other metrics, such as a normalized Levenshtein distance.

While I have some fringe ideas I'd like to test in the future, this is similar to the outlier identification above, but it "brute-forcely" looks for similar glosses. It builds its model from all concept lists in Concepticon and, as usual, the higher the score the better.

Here querying for a gloss that is found and is common, `father`:

```
In [4]: align_concept.get_candidates('father', model)
Out[4]: 
[('FATHER', 0.0),
 ('GATHER', -0.1823215567939546),
 ('MOTHER', -0.1823215567939546),
 ('FEATHER', -0.24275390789085033),
 ('WEATHER', -0.262364264467491),
 ('LEATHER', -0.262364264467491),
 ('STEPFATHER', -0.2876820724517809)]
```

Here a gloss with an obvious typo, `fath er`:

```
In [6]: align_concept.get_candidates('fath er', model)
Out[6]: 
[('FATHER', -0.3739360401245959),
 ('FATE', -0.5534555457414024),
 ('OATH', -0.6061358035703156),
 ('PATH', -0.6061358035703156),
 ('FEATHER', -0.6263814842476839),
 ('STEPFATHER', -0.6422000267192753),
 ('FAT (ORGANIC SUBSTANCE)', -0.8209805520698303),
 ('', -1.2992829841302609),
 ('BATHE', -1.2992829841302609),
 ('FAT (OBESE)', -2.238046571856475),
 ('FAT (OF MEAT AND FOOD)', -3.6243409329763647),
 ('FAT (FROM ANIMALS)', -3.6243409329763647)]
```

Here for a gloss with some annotation, `father (parent)`:

```
In [7]: align_concept.get_candidates('father (parent)', model)
Out[7]: 
[('GRANDFATHER', -0.44086917518374763),
 ('FATHER', -0.5518636946674401),
 ('PATERNAL GRANDFATHER', -0.5907905570055175),
 ('PARENTS', -0.6061358035703156),
 ('FEATHER', -0.6132228300889973)]
```

Here for an (invented) hyponym, `Brazilian coconut`:

```
In [8]: align_concept.get_candidates('Brazilian coconut', model)
Out[8]: 
[('GREEN COCONUT', -0.5046905774700834),
 ('COCONUT', -0.5389965007326869),
 ('RIPE COCONUT', -0.5869550144787102),
 ('COCONUT TREE', -0.7181371467124557),
 ('COCONUT SHELL LADLE', -0.9061893782153954),
 ('COCOA BEAN', -0.9812886007702626),
 ('', -0.9932517730102834),
 ('CRUCIAN CARP', -1.0451003804157877)]
```

And, at last, for a modern concept not in Concepticon, `keyboard`:

```
In [9]: align_concept.get_candidates('keyboard', model)
Out[9]: 
[('BOARD', -0.262364264467491),
 ('BOAR', -0.40546510810816444),
 ('', -0.4700036292457356),
 ('KEY', -0.6061358035703156),
 ('YARD', -0.6404669227310321),
 ('HARD', -0.6931471805599453),
 ('RUDDER', -1.0761394328160512),
 ('PADDLE', -1.992430164690206)]
```

As a rule of thumb, it seems that if the score of the best candidate is less than -0.4/-0.5, there are no good candidates.

**Comment by chrzyki** [comment]
_From @LinguList on November 28, 2017 6:56_

Yes, this goes along the lines for our concept mapping, although I think that our current mapping algorithm still has many advantages over this, as it looks at the problem from a different perspective. But let's call it "concept mapping", not "cognate mapping". I have few points to remark:

1 do you query against all sequences? If so, this will blow the search space and be inefficient with normal computers, even if linguistic strings are small. For that reasons, we use the workaround with dictionaries in our concept mapping approach.
2 do you only use the 3000 concepts in concepticon? Our mapping algorithm does more, which is also important to consider, as it takes all real concept-labels in the data and stores their links to concept_ids. As a result, we can even link "eggplant" to "aubergine", as it is a valid link. And switching to another language is trivial, as we just change the underlying data from our linked concept list which have a given language. For reference, check out our [lookup-lists](https://github.com/clld/concepticon-data/tree/master/pyconcepticon/data).
3 It is obviously not only surface similarity for glosses, but we also need:
  - a good stemming algorithm (currently poor in our implementation, but information in brackets is definitely not what we need most of the time)
  - a way to detect POS information (which people often use in English glosses)
  - a smart way to avoid all-to-all comparison when doing a lookup

I have been devoting quite some time to thinking how our algorithm could be improved, and here's what I think:

1 refine the stemming (also for different languages, includes POS detection)
2 refine the similarity decisions by the glossing algorithm
3 add a fuzzy component to the lookup procedure, e.g., by
  - systematically transforming a given lookup word (father -> fther, faher, etc.), which is equivalent to edit distance one, but stored in memory, as there's no need to do this on the fly
  - decide to go for more complex search (all-to-all) if no valid matches are found by the normal mapping procedure

In general, what is important, although I know that speed is not our major concern, is that we avoid n-to-n comparisons, but also don't restrict mapping to fuzzy sounds, as we have cases like "aubergine" and "eggplant" where we have no similarity at all. Plus, we need to use the knowledge we have assembled by linking already. One could even make a neural network out of our data, but I don't have the expertise for that.

**Comment by chrzyki** [comment]
_From @tresoldi on November 28, 2017 11:36_

Funny you mention the "egg plant"/"aubergine" case, as it is the one that bugged me (I almost submitted "eggplant" as a new concept, luckily I grepped the file for the scientific name first).

In short, I took a quick look at the current algorithm and decided to try something intentionally different, and focusing only on the glosses -- but as I said, please consider it just an idea shared among peers. On detail:

- Yes, it is a dumb query against all sequences. It is a computational waste, but at least it is not that expansive (it is actually quick for what it does); among the ideas I am playing with for a "smart way to avoid all-to-all comparisons" (also because I wish to use the concept description for the search, and because this could generalized to other column relationships) are standard NLP approaches as word vectors and some kind of locality-sensitive hashing for a (sub)string fingerprint (I could build on a method I once used, in short it is a crafted bloom filter on steroids -- once trained the search is linear, but it is still quite expensive from a computer science point of view). Both could be used to feed a neural net if you wanted, but I don't see much benefit in it (there are so few datapoints that either we don't converge or we overfit); maybe a random forest could help, but I'm just guessing. One of the advantages of this would be that it could be used to look for duplicate/similar concepts.

- I am only using the concepts in this gist, but I also quickly tried querying wordnet for the gloss, getting the description, and doing the brute-force searching among Concepticon descriptions (which is why I wanted to implement the word-vector/LSH, but would not have time for it). This actually worked well for the eggplant case and could be a fallback solution -- it is definitely an idea to consider.

- A long time ago I worked on a Portuguese stemmer, adapting the (English) Porter in NLTK ( http://www.nltk.org/howto/portuguese_en.html ); I could probably make it independent from NLTK if you want, but honestly my experience with NLP leads me to a Google approach of doing it purely with statistics (I am not really a fan of this approach, and I know you are even less and with good reason, but a "linguistic" solution devours your time...).

- the detection of POS information is the reason I wanted to use a string comparison function: I trained it with Concepticon concept lists, which are already somewhat normalized, but if it were trained on the raw dataset it should easily catch things like `(n)` or `-V$`.
